sympy/polys/matrices/domainscalar.py

```diff
@@ -0,0 +1,86 @@
+"""
+
+Module for the DomainScalar class.
+
+A DomainScalar represents an element which is in a particular
+Domain. The idea is that the DomainScalar class provides the
+convenience routines for unifying elements with different domains.
+
+It assists in Scalar Multiplication and getitem for DomainMatrix.
+
+"""
+from ..constructor import construct_domain
+
+
+class DomainScalar:
+    r"""
+    docstring
+    """
+
+    def __init__(self, element, domain):
+        self.element = element
+        self.domain = domain
+
+    @classmethod
+    def new(cls, element, domain):
+        return cls(element, domain)
+
+    def __repr__(self):
+        return repr(self.element)
+
+    @classmethod
+    def from_sympy(cls, expr):
+        [domain, [element]] = construct_domain([expr])
+        return cls.new(element, domain)
+
+    def to_domain(self, domain):
+        element = domain.convert_from(self.element, self.domain)
+        return self.new(element, domain)
+
+    def unify(self, other):
+        domain = self.domain.unify(other.domain)
+        return self.to_domain(domain), other.to_domain(domain)
+
+    def __add__(self, other):
+        if not isinstance(other, DomainScalar):
+            return NotImplemented
+        self, other = self.unify(other)
+        return self.new(self.element + other.element, self.domain)
+
+    def __sub__(self, other):
+        if not isinstance(other, DomainScalar):
+            return NotImplemented
+        self, other = self.unify(other)
+        return self.new(self.element - other.element, self.domain)
+
+    def __mul__(self, other):
+        if not isinstance(other, DomainScalar):
+            return NotImplemented
+        self, other = self.unify(other)
+        return self.new(self.element * other.element, self.domain)
+
+    def __floordiv__(self, other):
+        if not isinstance(other, DomainScalar):
+            return NotImplemented
+        self, other = self.unify(other)
+        K = self.domain
+        if not K.is_Field:
+            K = K.get_field()
+        return self.new(K.quo(self.element, other.element), self.domain)
+
+    def __pow__(self, n):
+        if not isinstance(n, int):
+            return NotImplemented
+        if self.element == 0 and n < 0:
+            raise ZeroDivisionError
+        return self.new(self.element**n, self.domain)
```

Comment:
I don't think any checking is needed here. This can just pass through to `self.element**n`.

Reply:
But then the cases like `self.element` being zero and `n` being negative will also pass, moreover n not being integer will also go through.

Reply:
We should check that `n` is `int` but I don't think we need to check for zero division because element will do that:
```python
In [15]: ZZ(0) ** -1                                                                                                              
---------------------------------------------------------------------------
ZeroDivisionError                         Traceback (most recent call last)
<ipython-input-15-d302bfe958af> in <module>
----> 1 ZZ(0) ** -1

ZeroDivisionError: 0.0 cannot be raised to a negative power
```

Reply:
Sure, thanks.